File: ActorClient/ActorClient/CocoaStorage.m
//
//  CocoaStorage.m
//  ActorModel
//
//  Created by Антон Буков on 17.02.15.
//  Copyright (c) 2015 Anton Bukov. All rights reserved.
//

#import "J2ObjC_source.h"
#import "im/actor/model/entity/Peer.h"
#import "im/actor/model/entity/User.h"
#import "im/actor/model/entity/Group.h"
#import "im/actor/model/entity/Dialog.h"
#import "im/actor/model/entity/Contact.h"
#import "im/actor/model/entity/Message.h"
#import "im/actor/model/modules/messages/entity/PendingMessagesStorage.h"
#import "UserDefaultsPreferencesStorage.h"
#import "CoreDataKeyValueStorage.h"
#import "CoreDataListEngine.h"
#import "ZonedCoreDataListEngine.h"
#import "CocoaStorage.h"

#import "AACDDialog.h"
#import "AACDDownload.h"
#import "AACDMessage.h"
#import "AACDUser.h"
#import "AACDGroup.h"
#import "AACDContact.h"

@implementation CocoaStorage

- (id<AMPreferencesStorage>)createPreferencesStorage
{
    return [[UserDefaultsPreferencesStorage alloc] init];
}

- (id<AMKeyValueStorage>)createUsersEngine
{
    return [[CoreDataKeyValueStorage alloc] initWithMOS:[AACDUser class]];
}

- (id<AMKeyValueStorage>)createGroupsEngine
{
    return [[CoreDataKeyValueStorage alloc] initWithMOS:[AACDGroup class]];
}

- (id<AMKeyValueStorage>)createDownloadsEngine
{
    return [[CoreDataKeyValueStorage alloc] initWithMOS:[AACDDownload class]];
}

- (id<AMListEngine>)createContactsEngine
{
    return [[CoreDataListEngine alloc] initWithMOS:[AACDContact class] serializer:^NSData *(AMContact *object) {
        return object.toByteArray.toNSData;
    } deserializer:^AMContact *(NSData *data) {
        IOSByteArray *byteArray = [IOSByteArray arrayWithBytes:data.bytes count:data.length];
        return byteArray.length ? [AMContact fromBytesWithByteArray:byteArray] : nil;
    }];
}

- (id<AMListEngine>)createDialogsEngine
{
    return [[CoreDataListEngine alloc] initWithMOS:[AACDDialog class] serializer:^NSData *(AMDialog *object) {
        return object.toByteArray.toNSData;
    } deserializer:^AMDialog *(NSData *data) {
        IOSByteArray *byteArray = [IOSByteArray arrayWithBytes:data.bytes count:data.length];
        return byteArray.length ? [AMDialog fromBytesWithByteArray:byteArray] : nil;
    }];
}

- (id<AMListEngine>)createMessagesEngineWithAMPeer:(AMPeer *)peer
{
    return [[ZonedCoreDataListEngine alloc] initWithMOS:[AACDMessage class] zone_id:peer.getPeerId serializer:^NSData *(AMMessage *object) {
        return object.toByteArray.toNSData;
    } deserializer:^AMMessage *(NSData *data) {
        IOSByteArray *byteArray = [IOSByteArray arrayWithBytes:[data bytes] count:[data length]];
        return byteArray.length ? [AMMessage fromBytesWithByteArray:byteArray] : nil;
    }];
}

@end

File: ActorClient/ActorClient/CocoaStorage.swift
//
//  CocoaStorage.swift
//  ActorClient
//
//  Created by Stepan Korshakov on 14.03.15.
//  Copyright (c) 2015 Actor LLC. All rights reserved.
//

import Foundation

@objc class CocoaStorage : NSObject, AMStorage {
    
    let dbPath: String;
    init(dbPath: String) {
        self.dbPath = dbPath;
    }
    
    func createPreferencesStorage() -> AMPreferencesStorage! {
        return UserDefaultsPreferencesStorage();
    }
    
    func createDownloadsEngine() -> AMKeyValueStorage! {
        return FMDBKeyValue(databasePath: dbPath, tableName: "downloads")
    }
    
    func createGroupsEngine() -> AMKeyValueStorage! {
        return FMDBKeyValue(databasePath: dbPath, tableName: "groups")
    }
    
    func createUsersEngine() -> AMKeyValueStorage! {
        return FMDBKeyValue(databasePath: dbPath, tableName: "users")
    }
    
    func createContactsEngine() -> AMListEngine! {
        return CoreDataListEngine(MOS: AACDContact.self, serializer: { (sourceItem) -> NSData! in
            return (sourceItem as! AMContact).toByteArray().toNSData();
            }, deserializer: { (data) -> AMListEngineItem! in
                if (data == nil) {
                    return nil
                }
                return AMContact.fromBytesWithByteArray(data.toJavaBytes());
        })
    }
    
    func createDialogsEngine() -> AMListEngine! {
        return CoreDataListEngine(MOS: AACDDialog.self, serializer: { (sourceItem) -> NSData! in
            return (sourceItem as! AMDialog).toByteArray().toNSData();
        }, deserializer: { (data) -> AMListEngineItem! in
            if (data == nil) {
                return nil
            }
            return AMDialog.fromBytesWithByteArray(data.toJavaBytes());
        })
    }
    
    func createMessagesEngineWithAMPeer(peer: AMPeer!) -> AMListEngine! {
        return ZonedCoreDataListEngine(MOS: AACDMessage.self, zone_id: peer.getPeerId(), serializer: { (sourceItem) -> NSData! in
            return (sourceItem as! AMMessage).toByteArray().toNSData();
        }, deserializer: { (data) -> AMListEngineItem! in
            if (data == nil) {
                return nil
            }
            return AMMessage.fromBytesWithByteArray(data.toJavaBytes());
        });
    }
}